pkg/networkmanager/firewall.jsx

```diff
@@ -171,6 +193,136 @@ class AddServicesBody extends React.Component {
         });
     }
 
+    parse_services(content) {
+        if (!content) {
+            console.warn("Couldn't read /etc/services");
+            return [ ];
+        }
+
+        var lines = content.split('\n');
+        var ret = {};
+
+        for (var i = 0; i < lines.length; i++) {
+            if (!lines[i] || lines[i].startsWith("#"))
+                continue;
+            var m = lines[i].match(/^(\S+)\s+(\d+)\/(\S+).*?(#(.*))?$/);
+            var new_port = { name: m[1], port: m[2], type: [m[3]] };
+            if (m.length > 5 && m[5])
+                new_port['description'] = m[5].trim();
+            if (ret[m[1]])
+                ret[m[1]]['type'].push(new_port['type'][0]);
+            else
+                ret[m[1]] = new_port;
+            if (ret[m[2]])
+                ret[m[2]]['type'].push(new_port['type'][0]);
+            else
+                ret[m[2]] = new_port;
+        }
+        return ret;
+    }
+
+    setName(event) {
+        this.setState({
+            custom_name: event.target.value,
+            generate_name: false,
+        });
+    }
+
+    getName(port) {
+        let known = this.state.avail_services[port];
+        if (known)
+            return known['name'];
+        else
+            return port;
+    }
+
+    validate(event) {
+        let ports = event.target.value.split(',');
+        let error = "";
+        let targets = ['tcp', 'custom_tcp_ports', 'tcp_error', 'custom_tcp_value'];
+        if (event.target.id === "udp-ports")
+            targets = ['udp', 'custom_udp_ports', 'udp_error', 'custom_udp_value'];
+
+        let new_ports = [];
+        let self = this;
+        ports.forEach(function(port) {
+            port = port.trim();
+            if (!port)
+                return;
+            let p = self.state.avail_services[port.trim()];
+            if (!p) {
+                let num_p = Number(port);
+                if (isNaN(num_p))
+                    error = "Unknown service name";
+                else if (num_p <= 0 || num_p > 65535)
+                    error = "Not valid port number";
+                else
+                    new_ports.push(port);
+            } else if (p['type'].indexOf(targets[0]) < 0)
+                error = "Not matching port number and type";
```

Comment:
"Mismatching..", or "Port number and type do not match".

Is that an explicit requirement of firewalld? People make up all kinds of ad-hoc ports for their custom services, many of which don't have a `services` entry.

Reply:
This is the one thing I am not entirely sure about. If port is mentioned in /etc/services but not with the selected type if to mark it as error or not. If for example `http - 80 - tcp -  Web Server` would be in /etc/services but not with `udp` what should we do?
1) If 80 is written in UDP
2) If `http` is written in UDP

Number one should be accepted and number two ignored as unknown port name or number two accepted as well and present it as `Web Server`?

Reply:
Tricky, I don't think there's an obviously right or wrong answer. So perhaps *if* the port is a well-known one, enforce the type, otherwise (for arbitrary unknown ports), allow anything. We can always extend that later on, but this will stay on the safe side.